File: 2D Arrays (Matrix)/searching_in_sorted_array.js
/*Name : Abhinav kumar
Github username : Abhinavcode13
Repository name : data-structures-and-algorithms
Problem : Search in 2D sorted array in Javascript
Issue Number : #272
Problem statement : 

Explanation of the below Javascript code :

The function searchMatrix takes in a matrix (matrix) and a target integer (target) as parameters. It returns true if the target is found in the matrix and false otherwise.

The function first checks for edge cases where the matrix is empty or the rows are empty, and immediately returns false in such cases.

It then initializes variables for the number of rows (rows) and columns (cols) in the matrix, and sets the left and right indices for the binary search.

The binary search is performed using a while loop, with the left and right indices as the condition. Inside the loop, the middle index (mid) is calculated using Math.floor((left + right) / 2), and the corresponding row and column indices are derived from the mid index.

The element at the mid index is compared with the target, and based on the comparison, the search space is narrowed down by updating the left and right indices accordingly.

If the target is found, the function returns true. If the loop completes without finding the target, the function returns false.

The time complexity of this JavaScript solution is O(log(m * n)), as it performs a binary search on a list of size m * n.


*/

-------------------------------------------------------------------------//Javascript code begins here-----------------------------------------------------------------------


/**
 * Searches for a target integer in a matrix.
 * @param {number[][]} matrix - The matrix to search in.
 * @param {number} target - The target integer to search for.
 * @return {boolean} - True if the target is found, false otherwise.
 */
function searchMatrix(matrix, target) {
    // Check for empty matrix or empty rows
    if (!matrix || matrix.length === 0 || matrix[0].length === 0) {
        return false;
    }

    const rows = matrix.length;
    const cols = matrix[0].length;
    let left = 0;
    let right = rows * cols - 1;

    // Perform binary search on the matrix
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const row = Math.floor(mid / cols);
        const col = mid % cols;

        if (matrix[row][col] === target) {
            // Target found
            return true;
        } else if (matrix[row][col] < target) {
            // Target is in the right half of the matrix
            left = mid + 1;
        } else {
            // Target is in the left half of the matrix
            right = mid - 1;
        }
    }

    // Target not found
    return false;
}

File: 2D Arrays (Matrix)/searching_in_sorted_array.cpp
/* Name : Rajeev Kumar
Github username : Tonystart121
Repository name : data-structures-and-algorithms
Problem : Searching in 2D sorted array in C++
Issue Number : #273
Problem statement : Given a sorted matrix mat[n][m] and an element ‘x’. Find the position of x in the matrix if it is present, else print -1.

Sample testcases: 

Testcase 1 --> 

Input: number of rows(n) and column(m).Let n=3,m=3 and value to search x = 20 .
arr[n][m] =  { {1, 5, 9},
              {14, 20, 21},
               {30, 34, 43} }

Output: found at (1,2);

Testcase 2 -->
Input: number of rows(n) and column(m).Let n=3,m=4 and value to search x = 43
arr[n][m] = { {1, 5, 9, 11},
              {14, 20, 21, 26},
              {30, 34, 43, 50} }
Output: Found at (2,3);

Time Complexity = O(n+m)
Space Complexity = O(n+m)


Explanation:
This code asks the user to enter the number of rows and column in the array and element to find in the array, and then prompts them to enter each element of the array one at a time. Once the array is complete, the code applies the linear search/mapping algorithm to find the element within the array, and then prints the position of that element to the console.

Start at the top left corner of the matrix.
Compare the target element to the element at the current position.
If the target element is equal to the element at the current position, then return the current position.
If the target element is less than the element at the current position, then move down one row.
If the target element is greater than the element at the current position, then move right one column.
Repeat steps 2-5 until the target element is found or the entire matrix has been searched.
*/

// ----------------------------------------------------------------------------- code begins now!


#include<bits/stdc++.h>
using namespace std;

int main(){

    // enter array rows and column
    int n,m;
    cin>>n>>m;

    // taking input array.
    int arr[n][m];
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            cin>>arr[i][j];
        }
    }
   
    // taking input value to search.
    int key;
    cin>>key;

    // initializing rightmost element as current element.
    int cr_row =  0, cr_col=m-1;
    bool ans = false;
    while(cr_row<n and cr_col>=0){
        
        // if key==curr output its position
        if(arr[cr_row][cr_col]==key){
            cout<<cr_row<<" "<<cr_col;
            ans=true;
            break;
        }
       
        // if key greater than array elements, row increasing.
        else if(key>arr[cr_row][cr_col]){
            cr_row++;
        }

        // if key less than array elements, col decreasing. 
        else if(key<arr[cr_row][cr_col]){
            cr_col--;
        }
    }

    // if not found outputting -1.
    if(ans==false) cout<<"-1"<<endl;

    return 0;
}